app/services/order_workflow.rb

```diff
@@ -64,8 +64,8 @@ def after_transition_hook(options)
   def persist_all_payments
     order.payments.each do |payment|
       original_payment_state = payment.state
-      if original_payment_state != payment.reload.state
-        payment.update(state: original_payment_state)
+      if original_payment_state != Spree::Payment.find(payment.id).state
+        payment.reload.update(state: original_payment_state)
```

Comment:
I'm still a bit confused what's happening here. Do we need a comment explaining that? My theory is:

- During checkout the order is going to state `payment` and a payment object is created.
- The order should go from state `payment` to `complete` but on error the state machine stays in `payment`.
- The payment object's normal states are: `checkout`, `pending`, `processing`, `completed`
- Error states are: `void` and `invalid`.
- A failed checkout causes a database rollback which reverts the payment state from `void` or `invalid` to `checkout` or `pending` (I don't know which).
- Despite the rollback, our in-memory payment object still has the failed state.
- Checking for a state change in the database is a sign of a rollback and we override the database with our payment state.
- In theory, if something else updated the payment record at the same time, like a webhook, we could accidentally override that. There is a potential race condition here but we are most likely fine because of the checkout sequence and synchronisation.

Do you agree with this?

What else could happen? I guess a payment could complete but something else fails and we override the payment state to mark it as completed. That's a good thing.

I reckon that the payment should happen outside of the state machine transition to not be affected by the rollback. The state machine transition just verifies that there is enough money for the order. Out of scope here.

Reply:
I dont agree about you being confused :-)
I think what's happening is exactly what you have described :ok_hand: and that was what I thought was happening :+1: 
I used some of your description to add code comments. I think it's now better explained, thanks!